Add endpoint to webapi to generate DB join token

mcbattirola opened the pull request:
The purpose of this PR is to add a new endpoint that creates DB join tokens in the web API.

To [simplify DB manual steps](https://github.com/gravitational/teleport/issues/10340), we want to have a join token on the web UI, so the user doesn't have to generate it via `tctl tokens add --type=db`. 

There is already an endpoint to get a join token to apps and nodes, now we're adding it for databases.

Changes:

- new endpoint `/webapi/databases/token`
- rename `createScriptJoinToken` to `createJoinToken`, since it just returns a token, without relation to any script necessarily 
- `createJoinToken` now receives a list of roles to create tokens to, instead of always creating for app and node

Here is the PR on `gravitational/webapps` that consumes this endpoint: https://github.com/gravitational/webapps/pull/653


Part of #10340

mcbattirola committed:
Add endpoint to generate db join tokens

mcbattirola committed:
Merge branch 'master' of github.com:gravitational/teleport into matheus/simplify-manual-db-join

mcbattirola committed:
Rename `createScriptJoinToken` to `createJoinToken`

mcbattirola committed:
Merge branch 'master' into matheus/simplify-manual-db-join

mcbattirola committed:
Merge branch 'master' into matheus/simplify-manual-db-join

mcbattirola committed:
Merge branch 'master' into matheus/simplify-manual-db-join

mcbattirola commented on lib/web/apiserver.go:
```diff
@@ -360,6 +360,8 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
 	// Database access handlers.
 	h.GET("/webapi/sites/:site/databases", h.WithClusterAuth(h.clusterDatabasesGet))
+	// database token generation
+	h.POST("/webapi/databases/token", h.WithAuth(h.createDatabaseJoinTokenHandle))
```
Or should we make an endpoint to retrieve tokens and have the role as parameter?

`POST webapi/token?role=node,app,db`

codingllama replied:
Do you think we could move this to gRPC?

No objections about the combined endpoint on my part, if you think it's feasible.

mcbattirola replied:
It looks like the only package in the `webapps` project that uses gRPC is `teleterm`.

I've searched docs and asked around but looks like we have no plans to use gRPC outside of `teleterm`, so I think it's best to keep it as a JSON endpoint, what do you think?

mcbattirola replied:
Yes, I think it makes sense to do it.

But I'm thinking about merging it this way (so we can deliver the feature) and going back and refactoring the endpoints in a future issue if you are not against it.

Does that make sense?

codingllama replied:
Sure, just leave a TODO for yourself (`// TODO(mcbattirola): ...`).

codingllama commented on lib/web/join_tokens.go:
```diff
@@ -62,7 +62,22 @@ func (h *Handler) createScriptJoinTokenHandle(w http.ResponseWriter, r *http.Req
 		return nil, trace.Wrap(err)
 	}
 
-	return createScriptJoinToken(r.Context(), clt)
```
Rename createScriptJoinTokenHandle to something more sensible too?

codingllama commented on lib/web/join_tokens.go:
```diff
@@ -62,7 +62,22 @@ func (h *Handler) createScriptJoinTokenHandle(w http.ResponseWriter, r *http.Req
+	return createJoinToken(r.Context(), clt, roles)
+}
+
+func (h *Handler) createDatabaseJoinTokenHandle(w http.ResponseWriter, r *http.Request, params httprouter.Params, ctx *SessionContext) (interface{}, error) {
```
Please create a test for the new endpoint. Let's exercise it via a client instead of calling private methods (which I see some of the existing tests too).

mcbattirola replied:
Sure, I added it in this commit: https://github.com/gravitational/teleport/pull/10914/commits/103a89403f43bf02c6d99553bf7940e99a3de834

I'm calling the endpoint and checking that a valid token with role `types.RoleDatabase` is created.

I had to add permission to create tokens to this [method that creates an admin](https://github.com/gravitational/teleport/blob/103a89403f43bf02c6d99553bf7940e99a3de834/lib/services/role.go#L124) user for testing. 
Since this is only used on tests and all tests are passing, I think this is ok.

codingllama replied:
> Note that I had to add permission to create tokens to this [method that creates an admin](https://github.com/gravitational/teleport/blob/103a89403f43bf02c6d99553bf7940e99a3de834/lib/services/role.go#L124) user for testing.

It seems alright, but if you need specific users/roles in the future, consider creating them in the test body.

codingllama reviewed (commented):
Sorry about the delay.

mcbattirola committed:
Rename token creation handler

codingllama commented on lib/web/apiserver.go:
```diff
@@ -352,7 +352,8 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
 
 	// add Node token generation
-	h.POST("/webapi/nodes/token", h.WithAuth(h.createScriptJoinTokenHandle))
+	h.POST("/webapi/nodes/token", h.WithAuth(h.createNodeTokenHandle))
```
Much better, thanks.

mcbattirola committed:
Add TODO to refactor endpoints

mcbattirola committed:
Merge branch 'master' into matheus/simplify-manual-db-join

mcbattirola committed:
Add test for `/webapi/databases/token`

mcbattirola committed:
Merge branch 'matheus/simplify-manual-db-join' of github.com:gravitational/teleport into matheus/simplify-manual-db-join

mcbattirola committed:
Merge branch 'master' into matheus/simplify-manual-db-join

mcbattirola committed:
Merge branch 'master' into matheus/simplify-manual-db-join

mcbattirola committed:
Merge branch 'master' of github.com:gravitational/teleport into matheus/simplify-manual-db-join

mcbattirola committed:
Merge branch 'master' of github.com:gravitational/teleport into matheus/simplify-manual-db-join

mcbattirola committed:
Add generic join token endpoint

mcbattirola committed:
Merge branch 'matheus/simplify-manual-db-join' of github.com:gravitational/teleport into matheus/simplify-manual-db-join

mcbattirola committed:
Add logs to errs

mcbattirola commented:
@codingllama @zmb3
Hey folks,
I've added in [this commit](https://github.com/gravitational/teleport/pull/10914/commits/476b8af4ee817492adfe1800396dfc678b450a4d) the new endpoint to token generation and removed `createDatabaseJoinTokenHandle`.

PTAL, thanks

mcbattirola committed:
Fix version to delete `/nodes/token` endpoint

mcbattirola commented on lib/web/apiserver.go:
```diff
@@ -351,8 +351,13 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
+
 	// add Node token generation
-	h.POST("/webapi/nodes/token", h.WithAuth(h.createScriptJoinTokenHandle))
+	// DELETE IN 10.0. Deprecated, use /webapi/token for generating tokens of any role.
```
Fixed, thanks

mcbattirola committed:
Merge branch 'master' into matheus/simplify-manual-db-join
